### PR TITLE
Make Plugins Proxies after transfering ownership

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Based on @carmocca s finding, this PR makes the plugins a proxy in the accelerator connector after moving the ownership to the accelerator.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -496,7 +499,8 @@ def resolve_training_type_plugin(self, training_type: TrainingTypePlugin) -> Tra
                 training_type.num_processes = len(self.parallel_devices)
 
         if hasattr(training_type, 'cluster_environment') and getattr(training_type, 'cluster_environment') is None:
-            training_type.cluster_environment = self.select_cluster_environment()
+            training_type.cluster_environment = self.cluster_environment
+            self._cluster_environment = proxy(self.cluster_environment)
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:503]:
```suggestion
            # transfer ownership of the cluster environment to the training type
            training_type.cluster_environment = self.cluster_environment
            self._cluster_environment = proxy(self.cluster_environment)
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -529,10 +533,14 @@ def select_accelerator(self) -> Accelerator:
             acc_cls = CPUAccelerator
         # as precision_plugin is dependent on training_type_plugin, make sure
         # that we first select training_type_plugin, then precision_plugin
-        return acc_cls(
+        accelerator = acc_cls(
             training_type_plugin=self.training_type_plugin,
             precision_plugin=self.precision_plugin,
         )
+        self._training_type_plugin = proxy(self.training_type_plugin)
+        self._precision_plugin = proxy(self.precision_plugin)
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:541]:
```suggestion
        # transfer ownership of the plugins to the accelerator
        self._training_type_plugin = proxy(self.training_type_plugin)
        self._precision_plugin = proxy(self.precision_plugin)
```

**Review by carmocca — APPROVED** [review]
Thanks for taking care of this!

**Comment by carmocca** [comment]
Should we add a test asserting the refcount of these?

**Comment by justusschock** [comment]
@carmocca These tests could be useful but I actually wouldn't add them here. I would add them in a separate PR for more objects then just these (which ones we still have to discuss ofc). Also thinking to do it for the model, devices, connectors and many more.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -529,10 +534,15 @@ def select_accelerator(self) -> Accelerator:
             acc_cls = CPUAccelerator
         # as precision_plugin is dependent on training_type_plugin, make sure
         # that we first select training_type_plugin, then precision_plugin
-        return acc_cls(
+        accelerator = acc_cls(
             training_type_plugin=self.training_type_plugin,
             precision_plugin=self.precision_plugin,
         )
+        # transfer ownership of the plugins to the accelerator
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:541]: in what sense do the lines below transfer ownership? is the comment meant to go above the accelerator creation?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -244,6 +245,8 @@ def training_type_plugin(self) -> TrainingTypePlugin:
 
     @property
     def cluster_environment(self) -> ClusterEnvironment:
+        if self._cluster_environment is None:
+            self._cluster_environment = self.select_cluster_environment()
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:249]: previously it was called directly so I introduced this to make this a proxy as well.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -529,10 +534,15 @@ def select_accelerator(self) -> Accelerator:
...
         # as precision_plugin is dependent on training_type_plugin, make sure
         # that we first select training_type_plugin, then precision_plugin
-        return acc_cls(
+        accelerator = acc_cls(
             training_type_plugin=self.training_type_plugin,
             precision_plugin=self.precision_plugin,
         )
+        # transfer ownership of the plugins to the accelerator
+        self._training_type_plugin = proxy(self.training_type_plugin)
+        self._precision_plugin = proxy(self.precision_plugin)
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:543]: No we can't since the first have to be given to the accelerator (so that we have a hard reference somewhere). Otherwise we end up with no reference at all which may result in these objects being deleted too early.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:541]: before the accelerator was created before, the ownership of those plugins was in the connector (by potentially being the only class to hold a strong reference to them). Now that the accelerator has them, it should be the only place with a strong ref.

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]
